cpp/src/arrow/filesystem/s3fs.cc

```diff
@@ -3389,6 +3389,9 @@ struct AwsInstance {
         ARROW_LOG(WARNING)
             << " arrow::fs::FinalizeS3 was not called even though S3 was initialized.  "
                "This could lead to a segmentation fault at exit";
+        auto leaked_ptr = new std::shared_ptr<S3ClientFinalizer>(GetClientFinalizer());
```

Comment:
it took me a while to understand this wasn't like a make_shared. you could use `auto *` here.